service/matching/matching_engine_test.go

```diff
@@ -397,6 +397,81 @@ func (s *matchingEngineSuite) TestOnlyUnloadMatchingInstance() {
 		"Unload call with matching incarnation should have caused unload")
 }
 
+func (s *matchingEngineSuite) TestFailAddTaskWithHistoryError() {
+	namespaceID := namespace.ID(uuid.New())
+	identity := "identity"
+
+	stickyTaskQueue := &taskqueuepb.TaskQueue{Name: "STQ", Kind: enumspb.TASK_QUEUE_KIND_STICKY}
+
+	s.matchingEngine.config.LongPollExpirationInterval = dynamicconfig.GetDurationPropertyFnFilteredByTaskQueue(5 * time.Second)
+
+	runID := uuid.NewRandom().String()
+	workflowID := "workflow1"
+	execution := &commonpb.WorkflowExecution{RunId: runID, WorkflowId: workflowID}
+	scheduledEventID := int64(0)
+
+	expectedError := serviceerror.NewInternal("nothing to start")
+	s.mockHistoryClient.EXPECT().RecordWorkflowTaskStarted(gomock.Any(), gomock.Any(), gomock.Any()).Return(
+		nil, expectedError)
+
+	addRequest := matchingservice.AddWorkflowTaskRequest{
+		NamespaceId:            namespaceID.String(),
+		Execution:              execution,
+		ScheduledEventId:       scheduledEventID,
+		TaskQueue:              stickyTaskQueue,
+		ScheduleToStartTimeout: timestamp.DurationFromSeconds(100),
+	}
+
+	pollRequest := matchingservice.PollWorkflowTaskQueueRequest{
+		NamespaceId: namespaceID.String(),
+		PollRequest: &workflowservice.PollWorkflowTaskQueueRequest{
+			TaskQueue: stickyTaskQueue,
+			Identity:  identity,
+		},
+	}
+
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+
+	go func() {
+		wg.Done()
+		resp, _ := s.matchingEngine.PollWorkflowTaskQueue(context.Background(), &pollRequest, metrics.NoopMetricsHandler)
+		if resp != emptyPollWorkflowTaskQueueResponse {
+			s.logger.Info("Polling task got empty response")
+		}
+	}()
+	wg.Wait()
+	time.Sleep(100 * time.Millisecond) //nolint:forbidigo
+	partition, err := tqid.PartitionFromProto(addRequest.TaskQueue, addRequest.NamespaceId, enumspb.TASK_QUEUE_TYPE_WORKFLOW)
+	s.NoError(err)
+
+	pm, err := s.matchingEngine.getTaskQueuePartitionManager(context.Background(), partition, true, loadCausePoll)
+	s.NoError(err)
+	s.NotEmpty(pm)
+
+	hasPoller := pm.HasAnyPollerAfter(time.Now().Add(-noPollerThreshold))
+	s.True(hasPoller)
```

Comment:
using s.matchingEngine.DescribeTaskQueuePartition and check the returned pollers count may be slightly easier.